# Zenodo Data Initial Model Training Multimodal Drone Detection
**Ethan Mauger**

The following notebook details training steps and evaluation metrics on the three Zenodo Drone Detection Datasets. This notebook was ran in Google Colab to utilize the free T4 GPU that Google provides access to. To access this notebook in Google Colab, go to https://colab.research.google.com/.

There, select the GitHub tab and input your GitHub username. Scroll down to the azizs1/multimodal-drone-detection repo and click on the main branch, and select this notebook. Click on the Runtime header at the top and then click on Change runtime type, and ensure that the T4 GPU button is clicked after clicking save.

Next, you must copy all of the data from your local machine to your google drive. The paths are currently set up for the datasets folder to be at the top when you click on My Drive in Google Drive. Once all of the datasets are copied with the same structure that they appear as in VS Code, run the next cell and authorize Colab to access your Google Drive.

After installing YOLO libraries, you can run the code as follows. This code operates on the yolov8n.pt model, which is YOLO's most efficient model, and currently only runs 50 epochs.

Metrics and analysis will be displayed below.

In [1]:
#Import datasets from Google Drive
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [2]:
#Install YOLO Libraries
!pip install ultralytics
from ultralytics import YOLO

In [3]:
#Load Model (options include yolov8n.pt (smallest), yolov8s.pt, yolov8m.pt, yolov8l.pt, yolov8x.pt).
model1 = YOLO('yolov8n.pt')

#Copy data to local Colab storage for faster caching.
!cp -r "/content/drive/MyDrive/datasets/zenodo_visual_no_augmentation" /content/

#Train visual_no_augmentation model.
model1.train(
    data='/content/zenodo_visual_no_augmentation/data.yaml',
    epochs=50,
    project='/content/drive/MyDrive/runs',
    name='visual_no_augmentation_baseline'
)

Ultralytics 8.4.19 🚀 Python-3.12.12 torch-2.10.0+cu128 CUDA:0 (Tesla T4, 14913MiB)
engine/trainer: agnostic_nms=False, amp=True, angle=1.0, augment=False, auto_augment=randaugment, batch=16, bgr=0.0, box=7.5, cache=False, cfg=None, classes=None, close_mosaic=10, cls=0.5, compile=False, conf=None, copy_paste=0.0, copy_paste_mode=flip, cos_lr=False, cutmix=0.0, data=/content/zenodo_visual_no_augmentation/data.yaml, degrees=0.0, deterministic=True, device=None, dfl=1.5, dnn=False, dropout=0.0, dynamic=False, embed=None, end2end=None, epochs=50, erasing=0.4, exist_ok=False, fliplr=0.5, flipud=0.0, format=torchscript, fraction=1.0, freeze=None, half=False, hsv_h=0.015, hsv_s=0.7, hsv_v=0.4, imgsz=640, int8=False, iou=0.7, keras=False, kobj=1.0, line_width=None, lr0=0.01, lrf=0.01, mask_ratio=4, max_det=300, mixup=0.0, mode=train, model=yolov8n.pt, momentum=0.937, mosaic=1.0, multi_scale=0.0, name=visual_no_augmentation_baseline2, nbs=64, nms=False, opset=None, optimize=False, optimizer=auto

ultralytics.utils.metrics.DetMetrics object with attributes:

ap_class_index: array([0])
box: ultralytics.utils.metrics.Metric object
confusion_matrix: <ultralytics.utils.metrics.ConfusionMatrix object at 0x7d375b477bc0>
curves: ['Precision-Recall(B)', 'F1-Confidence(B)', 'Precision-Confidence(B)', 'Recall-Confidence(B)']
curves_results: [[array([          0,    0.001001,    0.002002,    0.003003,    0.004004,    0.005005,    0.006006,    0.007007,    0.008008,    0.009009,     0.01001,    0.011011,    0.012012,    0.013013,    0.014014,    0.015015,    0.016016,    0.017017,    0.018018,    0.019019,     0.02002,    0.021021,    0.022022,    0.023023,
          0.024024,    0.025025,    0.026026,    0.027027,    0.028028,    0.029029,     0.03003,    0.031031,    0.032032,    0.033033,    0.034034,    0.035035,    0.036036,    0.037037,    0.038038,    0.039039,     0.04004,    0.041041,    0.042042,    0.043043,    0.044044,    0.045045,    0.046046,    0.047047,
          0.048048, 

In [4]:
#Load Model (options include yolov8n.pt (smallest), yolov8s.pt, yolov8m.pt, yolov8l.pt, yolov8x.pt).
model2 = YOLO('yolov8n.pt')

#Copy data to local Colab storage for faster caching.
!cp -r "/content/drive/MyDrive/datasets/zenodo_thermal_no_augmentation" /content/

#Train thermal_no_augmentation model.
model2.train(
    data='/content/zenodo_thermal_no_augmentation/data.yaml',
    epochs=50,
    project='/content/drive/MyDrive/runs',
    name='thermal_no_augmentation_baseline'
)

Ultralytics 8.4.19 🚀 Python-3.12.12 torch-2.10.0+cu128 CUDA:0 (Tesla T4, 14913MiB)
engine/trainer: agnostic_nms=False, amp=True, angle=1.0, augment=False, auto_augment=randaugment, batch=16, bgr=0.0, box=7.5, cache=False, cfg=None, classes=None, close_mosaic=10, cls=0.5, compile=False, conf=None, copy_paste=0.0, copy_paste_mode=flip, cos_lr=False, cutmix=0.0, data=/content/zenodo_thermal_no_augmentation/data.yaml, degrees=0.0, deterministic=True, device=None, dfl=1.5, dnn=False, dropout=0.0, dynamic=False, embed=None, end2end=None, epochs=50, erasing=0.4, exist_ok=False, fliplr=0.5, flipud=0.0, format=torchscript, fraction=1.0, freeze=None, half=False, hsv_h=0.015, hsv_s=0.7, hsv_v=0.4, imgsz=640, int8=False, iou=0.7, keras=False, kobj=1.0, line_width=None, lr0=0.01, lrf=0.01, mask_ratio=4, max_det=300, mixup=0.0, mode=train, model=yolov8n.pt, momentum=0.937, mosaic=1.0, multi_scale=0.0, name=thermal_no_augmentation_baseline, nbs=64, nms=False, opset=None, optimize=False, optimizer=aut

ultralytics.utils.metrics.DetMetrics object with attributes:

ap_class_index: array([0])
box: ultralytics.utils.metrics.Metric object
confusion_matrix: <ultralytics.utils.metrics.ConfusionMatrix object at 0x7d3746118e60>
curves: ['Precision-Recall(B)', 'F1-Confidence(B)', 'Precision-Confidence(B)', 'Recall-Confidence(B)']
curves_results: [[array([          0,    0.001001,    0.002002,    0.003003,    0.004004,    0.005005,    0.006006,    0.007007,    0.008008,    0.009009,     0.01001,    0.011011,    0.012012,    0.013013,    0.014014,    0.015015,    0.016016,    0.017017,    0.018018,    0.019019,     0.02002,    0.021021,    0.022022,    0.023023,
          0.024024,    0.025025,    0.026026,    0.027027,    0.028028,    0.029029,     0.03003,    0.031031,    0.032032,    0.033033,    0.034034,    0.035035,    0.036036,    0.037037,    0.038038,    0.039039,     0.04004,    0.041041,    0.042042,    0.043043,    0.044044,    0.045045,    0.046046,    0.047047,
          0.048048, 

In [7]:
#NOTE: Takes about 3 hours to train, so I waited to train this at baseline after Sprint 1

#Load Model (options include yolov8n.pt (smallest), yolov8s.pt, yolov8m.pt, yolov8l.pt, yolov8x.pt).
model3 = YOLO('yolov8n.pt')

#Copy data to local Colab storage for faster caching.
!cp -r "/content/drive/MyDrive/datasets/zenodo_thermal_augmented" /content/

#Train thermal_augmented model.
model3.train(
    data='/content/zenodo_thermal_augmented/data.yaml',
    epochs=50,
    project='/content/drive/MyDrive/runs',
    name='thermal_augmented_baseline'
)

^C
Ultralytics 8.4.19 🚀 Python-3.12.12 torch-2.10.0+cu128 CUDA:0 (Tesla T4, 14913MiB)
engine/trainer: agnostic_nms=False, amp=True, angle=1.0, augment=False, auto_augment=randaugment, batch=16, bgr=0.0, box=7.5, cache=False, cfg=None, classes=None, close_mosaic=10, cls=0.5, compile=False, conf=None, copy_paste=0.0, copy_paste_mode=flip, cos_lr=False, cutmix=0.0, data=/content/zenodo_thermal_augmented/data.yaml, degrees=0.0, deterministic=True, device=None, dfl=1.5, dnn=False, dropout=0.0, dynamic=False, embed=None, end2end=None, epochs=50, erasing=0.4, exist_ok=False, fliplr=0.5, flipud=0.0, format=torchscript, fraction=1.0, freeze=None, half=False, hsv_h=0.015, hsv_s=0.7, hsv_v=0.4, imgsz=640, int8=False, iou=0.7, keras=False, kobj=1.0, line_width=None, lr0=0.01, lrf=0.01, mask_ratio=4, max_det=300, mixup=0.0, mode=train, model=yolov8n.pt, momentum=0.937, mosaic=1.0, multi_scale=0.0, name=thermal_augmented_baseline2, nbs=64, nms=False, opset=None, optimize=False, optimizer=auto, overl

KeyboardInterrupt: 

In [11]:
#Load trained model and evaluate.
model1 = YOLO('/content/drive/MyDrive/runs/visual_no_augmentation_baseline/weights/best.pt')
model2 = YOLO('/content/drive/MyDrive/runs/thermal_no_augmentation_baseline/weights/best.pt')
#model3 = YOLO('/content/drive/MyDrive/runs/thermal_augmented_baseline/weights/best.pt')

#Validate on the test set.
metrics1 = model1.val(data='/content/zenodo_visual_no_augmentation/data.yaml', split='test', plots=True)
metrics2 = model2.val(data='/content/zenodo_thermal_no_augmentation/data.yaml', split='test', plots=True)
#metrics3 = model3.val(data='/content/zenodo_thermal_augmented/data.yaml', split='test', plots=True)

Ultralytics 8.4.19 🚀 Python-3.12.12 torch-2.10.0+cu128 CUDA:0 (Tesla T4, 14913MiB)
Model summary (fused): 73 layers, 3,005,843 parameters, 0 gradients, 8.1 GFLOPs
val: Fast image access ✅ (ping: 0.0±0.0 ms, read: 1056.0±194.7 MB/s, size: 62.6 KB)
val: Scanning /content/zenodo_visual_no_augmentation/test/labels.cache... 214 images, 0 backgrounds, 0 corrupt: 100% ━━━━━━━━━━━━ 214/214 56.1Mit/s 0.0s
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100% ━━━━━━━━━━━━ 14/14 2.4it/s 5.7s
                   all        214        216       0.93      0.919      0.945       0.56
Speed: 1.8ms preprocess, 4.8ms inference, 0.0ms loss, 5.2ms postprocess per image
Results saved to /content/runs/detect/val8
Ultralytics 8.4.19 🚀 Python-3.12.12 torch-2.10.0+cu128 CUDA:0 (Tesla T4, 14913MiB)
Model summary (fused): 73 layers, 3,005,843 parameters, 0 gradients, 8.1 GFLOPs
val: Fast image access ✅ (ping: 0.0±0.0 ms, read: 990.1±406.8 MB/s, size: 42.5 KB)
val: Scanning

## Basic Metrics

It is clear that both of the initial non-augmented visual and thermal models performed very will, with mAP50 (precision) at 94.5% and 97.9% respectively. Both had low false positives and false negatives indicated by the P and R statistics above. For more information and visuals on this information, check out the runs folder for the corresponding model.